# COURSE:   PGP [AI&ML]
## Learner :  Chaitanya Kumar Battula
## Topic : Machine Learning
## Project title:  Practice Project _Simple Linear Regression

In [1]:
import numpy as np
import pandas as pd

In [2]:
salaryData = pd.read_csv('Salary_Data.csv')

In [3]:
salaryData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 2 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   YearsExperience  30 non-null     float64
 1   Salary           30 non-null     float64
dtypes: float64(2)
memory usage: 608.0 bytes


In [4]:
salaryData.shape

(30, 2)

In [5]:
salaryData.head()

YearsExperience   Salary
0              1.1  39343.0
1              1.3  46205.0
2              1.5  37731.0
3              2.0  43525.0
4              2.2  39891.0

In [6]:
salaryData.isna().any()

YearsExperience    False
Salary             False
dtype: bool

In [8]:
salaryData.isnull().sum()

YearsExperience    0
Salary             0
dtype: int64

In [10]:
salaryData.describe().T

count          mean           std      min       25%  \
YearsExperience   30.0      5.313333      2.837888      1.1      3.20   
Salary            30.0  76003.000000  27414.429785  37731.0  56720.75   

                     50%        75%       max  
YearsExperience      4.7       7.70      10.5  
Salary           65237.0  100544.75  122391.0

The mean years of experience is 5.31 years

Rules for Regression:

1. Features and the target should be in ndarray.
2. Features should be in 2d and target should be 2d. 

In [11]:
features = salaryData.YearsExperience.values
target = salaryData.Salary.values

In [12]:
features.shape

(30,)

In [13]:
target.shape

(30,)

In [14]:
features = features.reshape(30,1)

In [51]:
features.reshape(-1,1)

array([[ 1.1],
       [ 1.3],
       [ 1.5],
       [ 2. ],
       [ 2.2],
       [ 2.9],
       [ 3. ],
       [ 3.2],
       [ 3.2],
       [ 3.7],
       [ 3.9],
       [ 4. ],
       [ 4. ],
       [ 4.1],
       [ 4.5],
       [ 4.9],
       [ 5.1],
       [ 5.3],
       [ 5.9],
       [ 6. ],
       [ 6.8],
       [ 7.1],
       [ 7.9],
       [ 8.2],
       [ 8.7],
       [ 9. ],
       [ 9.5],
       [ 9.6],
       [10.3],
       [10.5]])

In [52]:
target = target.reshape(-1,1)

In [53]:
target.shape

(30, 1)

In [54]:
features.shape

(30, 1)

In [55]:
type(features)

numpy.ndarray

In [56]:
type(target)

numpy.ndarray

In [57]:
from sklearn.model_selection import train_test_split

In [58]:
X_train,X_test,y_train,y_test = train_test_split(features,target,test_size=0.3,random_state=10)

In [59]:
X_train.shape

(21, 1)

In [60]:
X_test.shape

(9, 1)

In [61]:
y_train.shape

(21, 1)

In [62]:
y_test.shape

(9, 1)

In [63]:
X_train

array([[ 1.3],
       [ 6. ],
       [ 4.5],
       [ 5.9],
       [ 3. ],
       [ 4. ],
       [ 8.2],
       [ 8.7],
       [10.3],
       [ 7.9],
       [ 3.9],
       [ 9.5],
       [10.5],
       [ 3.2],
       [ 9. ],
       [ 5.1],
       [ 5.3],
       [ 1.1],
       [ 4.9],
       [ 2.2],
       [ 3.7]])

In [64]:
from sklearn.linear_model import LinearRegression

In [65]:
model = LinearRegression()

In [66]:
model.fit(X_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [67]:
y_preds = model.predict(X_test)

In [69]:
from sklearn.metrics import r2_score

In [70]:
r2_score(y_test,y_preds) # Testing Accuracy

0.9647278344670828

In [72]:
model.score(features,target) #Training Accuracy

0.9561671324383834

In [73]:
from sklearn.metrics import mean_squared_error

In [74]:
mean_squared_error(y_test,y_preds)

21713548.637118638

In [75]:
np.sqrt(mean_squared_error(y_test,y_preds))

4659.7798914882915

In [76]:
model.coef_

array([[9303.95933197]])

In [77]:
model.intercept_

array([27206.42890293])

y = 9303x + 27206

In [78]:
#Pre Deployment Test

In [80]:
yExp = float(input("Enter years of experience:"))
predSal = model.predict(np.array([[yExp]]))

print("The salary for {} years of experience is {}".format(yExp,predSal))

Enter years of experience: 80


The salary for 80.0 years of experience is [[771523.17546029]]
